scintillometry/io/psrfits_io.py

```diff
@@ -0,0 +1,488 @@
+    def __new__(cls, header_hdu, subint_hdu=None):
+        # Map Subint subclasses
+        mode = header_hdu.obs_mode
+        try:
+            cls = subint_map[mode]
+        except KeyError:
+            raise ValueError("'{}' is not a valid mode.".format(mode))
+        return super(SubintHDUBase, cls).__new__(cls)
+
+    def __init__(self, header_hdu, subint_hdu=None):
+        self.header_hdu = header_hdu
+        if subint_hdu is None:
+            super(SubintHDUBase, self).__init__(name='SUBINT')
+        else:
+            super(SubintHDUBase, self).__init__(header=subint_hdu.header,
+                                            data=subint_hdu.data)
+        self.verify()
+        self.offset = 0
+
+    def verify(self):
+        if self.header['EXTNAME'].strip() != "SUBINT":
+            raise ValueError("Input HDU is not a SUBINT type.")
+        # Check frequency
+        if 'DAT_FREQ' in self.columns.names:
+            freqs = u.Quantity(self.data['DAT_FREQ'],
+                               u.MHz, copy=False)
+            assert np.isclose(freqs, freqs[0],
+                              atol=np.finfo(float).eps * u.MHz).all(), \
+                "Frequencies are different within one subint rows."
+
+    @property
+    def mode(self):
+        return self.header_hdu.obs_mode
+
+    @property
+    def sample_rate(self):
+        raise NotImplementedError
+
+    @sample_rate.setter
+    def sample_rate(self, val):
+        raise NotImplementedError
+
+    @property
+    def start_time(self):
+        # NOTE should we get the start time for each raw, in case the time gaps
+        # in between the rows
+        file_start = self.header_hdu.start_time
+        return start_time
+
+    @start_time.setter
+    def start_time(self, time):
+        # NOTE this sets the start time of the HDU, not the file start time.
+        dt = (time - self.header_hdu.start_time).to(u.s)
+        center_off = dt + 1.0 / self.sample_rate * self.samples_per_frame / 2
+        if "OFFS_SUB" in self.columns.names:
+            first_off = self.data['OFFS_SUB'][0]
+            self.data['OFFS_SUB'] = (self.data['OFFS_SUB'] - first_off +
+                                     center_off)
+        else:
+            self._req_columns['OFFS_SUB']['value'] = center_off
+
+    @property
+    def nrow(self):
+        return int(self.header['NAXIS2'])
+
+    @property
+    def nchan(self):
+        return int(self.header['NCHAN'])
+
+    @property
+    def npol(self):
+        return int(self.header['NPOL'])
+
+    @property
+    def nbin(self):
+        return int(self.header['NBIN'])
+
+    @property
+    def samples_per_frame(self):
+        try:
+            return int(self.header['NSBLK'])
+        except:
+            return int(np.prod(self.data_shape)/(self.nrow * self.nchan *
+                                                 self.npol * self.nbin))
+
+    @samples_per_frame.setter
+    def samples_per_frame(self, val):
+        self.header['NSBLK'] = val
```

Comment:
Without checking consistency, I don't think this setter would be a good idea. It again suggests to write the setters when we need them, and when they will be properly tested!

Reply:
Move to the writing PR